<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="debug" default="true"/>
  <arg name="yolo_model" default="yolov8m-seg.pt"/>
  <arg name="input_topic" default="/image_jpeg/compressed"/>
  <arg name="result_topic" default="/yolo_result"/>
  <arg name="result_image_topic" default="/yolo_image"/>
  <arg name="conf_thres" default="0.5"/>
  <arg name="iou_thres" default="0.45"/>
  <arg name="max_det" default="300"/>
  <arg name="classes" default=""/>
  <arg name="tracker" default="bytetrack.yaml"/>
  <arg name="device" default="0"/>
  <arg name="result_conf" default="true"/>
  <arg name="result_line_width" default="1"/>
  <arg name="result_font_size" default="1"/>
  <arg name="result_font" default="Arial.ttf"/>
  <arg name="result_labels" default="true"/>
  <arg name="result_boxes" default="true"/>
  
  <!-- 라이다 관련 파라미터 -->
  <arg name="lidar_topic" default="/velodyne_points"/>
  <arg name="emergency_distance" default="1.0"/>  <!-- 긴급 정지 트리거 거리 (미터) -->
  <arg name="frame_id" default="ego_car"/>  <!-- 메인 프레임 ID -->
  
  <!-- 퓨전 모드 설정 -->
  <!-- 0: 기존 퓨전 방식, 1: 클러스터링 기반 퓨전, 2: 두 방식 모두 사용 -->
  <arg name="fusion_mode" default="2"/>
  
  <!-- 클러스터링 파라미터 - 객체 크기에 맞게 조정 -->
  <arg name="cluster_tolerance" default="0.7"/>   <!-- 점 간 최대 거리: 0.7m -->
  <arg name="min_cluster_size" default="15"/>    <!-- 최소 포인트 수: 15개 -->
  <arg name="max_cluster_size" default="5000"/>  <!-- 최대 포인트 수: 5000개 -->
  
  <!-- 캘리브레이션 파라미터 파일 로드 -->
  <rosparam command="load" file="$(find creamsooo_fusion)/config/calibration_params.yaml"/>
  
  <!-- YOLO 객체 감지 노드 -->
  <node name="tracker_node" pkg="ultralytics_ros" type="tracker_node.py" output="screen">
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="input_topic" value="$(arg input_topic)"/>
    <param name="result_topic" value="$(arg result_topic)"/>
    <param name="result_image_topic" value="$(arg result_image_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="result_conf" value="$(arg result_conf)"/>
    <param name="result_line_width" value="$(arg result_line_width)"/>
    <param name="result_font_size" value="$(arg result_font_size)"/>
    <param name="result_font" value="$(arg result_font)"/>
    <param name="result_labels" value="$(arg result_labels)"/>
    <param name="result_boxes" value="$(arg result_boxes)"/>
    <rosparam param="classes" subst_value="true">$(arg classes)</rosparam>
    <rosparam param="device" subst_value="true">$(arg device)</rosparam>
  </node>
  
  <!-- creamsooo_fusion 노드 -->
  <node name="creamsooo_fusion_node" pkg="creamsooo_fusion" type="creamsooo_fusion_node" output="screen">
    <param name="lidar_topic" value="$(arg lidar_topic)"/>
    <param name="yolo_result_topic" value="$(arg result_topic)"/>
    <param name="emergency_distance" value="$(arg emergency_distance)"/>
    <param name="frame_id" value="$(arg frame_id)"/>
    <param name="fusion_mode" value="$(arg fusion_mode)"/>
    <param name="cluster_tolerance" value="$(arg cluster_tolerance)"/>
    <param name="min_cluster_size" value="$(arg min_cluster_size)"/>
    <param name="max_cluster_size" value="$(arg max_cluster_size)"/>
  </node>
  
  <!-- RViz 표시 -->
  <node if="$(arg debug)" pkg="rviz" type="rviz" name="rviz" args="-d $(find creamsooo_fusion)/rviz/creamsooo_fusion.rviz"/>
</launch> 